Feature: Admin Screen

LouisJamesVasseur opened the pull request:
### What is this PR?

This PR allows administrators (a special type of User defined in a future PR), to have a clear view of all Reported Objects, as well as the list of all Reports that led to the addition of that Object (Parking/Review) into ReportedObjects.

### Why?

This brings the moderation of reports "full circle": they can now be productively used to remove unwanted objects from the app

### How?
 
Accessible from an Admin's Profile Screen, two Screens were introduced: the Admin Screen, where users can view all Reported Parkings and all Reported Reviews/Parkings by toggling the Sorting option. Users can then, by expanding and clicking on the corresponding card, view the list of all Reports that were made against that Object, and then choose to delete the object. This will delete the object from both the Reported Objects Repository and the Repository that Object originally comes from

<img width="356" alt="Screenshot 2024-11-28 at 03 00 29" src="https://github.com/user-attachments/assets/b503f5d2-132d-48cc-8ad0-3615b8f101f2">
<img width="386" alt="Screenshot 2024-11-28 at 03 04 41" src="https://github.com/user-attachments/assets/5bae5fee-051e-428f-b904-279870c2e076">
<img width="359" alt="Screenshot 2024-11-28 at 03 05 05" src="https://github.com/user-attachments/assets/4fb5b42f-ad37-43fd-81d2-0f6627a09b96">

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -332,40 +360,72 @@ class ParkingViewModel(
+                  reportUID = report.uid,
+                  nbOfTimesReported = selectedParking.nbReports + 1,
+                  nbOfTimesMaxSeverityReported =
+                      if (report.reason.severity == 3) selectedParking.nbMaxSeverityReports + 1
```
Missing MAX_SEVERITY instead of 3

matttsss commented on app/src/main/java/com/github/se/cyrcle/model/report/ReportedObjectRepository.kt:
```diff
@@ -1,5 +1,7 @@
 package com.github.se.cyrcle.model.report
 
+import com.google.firebase.firestore.DocumentReference
```
As mentioned above, this import should not exist in the Interface...

LouisJamesVasseur replied:
I will fix this, no worries

matttsss commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -16,7 +15,7 @@ const val NB_REPORTS_MAXSEVERITY_THRESH = 3
 
 class ReviewViewModel(
-    private val reviewRepository: ReviewRepository,
+    val reviewRepository: ReviewRepository,
```
Same comment, why is it now public?

LouisJamesVasseur replied:
Same response :)

matttsss commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -83,41 +100,81 @@ class ReviewViewModel(
+                      else selectedReview.nbMaxSeverityReports,
+                  userUID = user.public.userId,
+                  objectType = ReportedObjectType.REVIEW)
+          Log.d("AAAAAAAAAAAAAAA", selectedReview.nbReports.toString())
```
Nice `Log` tag :) But maybe a more relevant one would be nice

LouisJamesVasseur replied:
Oops, since I left it as draft I forgot to remove then!

matttsss commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -83,41 +100,81 @@ class ReviewViewModel(
+                  objectType = ReportedObjectType.REVIEW)
+          Log.d("AAAAAAAAAAAAAAA", selectedReview.nbReports.toString())
+
+          reportedObjectRepository.getDocumentReferenceByObjectUID(
```
Same as for the `ParkingViewModel`...

matttsss commented on app/src/main/java/com/github/se/cyrcle/ui/parkingDetails/ParkingDetailsAlertDialogShowImage.kt.kt:
```diff
@@ -0,0 +1,60 @@
+ */
+@Composable
+@OptIn(ExperimentalMaterial3Api::class)
+fun ParkingDetailsAlertDialogShowImage(onDismiss: () -> Unit, imageUrl: String) {
```
I'm confused, does this come from @sean-prz's PR? Did you rebase on main correctly?

LouisJamesVasseur replied:
Just noticed there was an issue with that... didn't really do anything different to normal. Oh well, unfortunately nothing can really be done about it now and it's adding the same code onto main so I'll leave it as is

matttsss commented on app/src/main/java/com/github/se/cyrcle/ui/report/ViewReportsScreen.kt:
```diff
@@ -0,0 +1,167 @@
+    parkingViewModel: ParkingViewModel,
+    reviewViewModel: ReviewViewModel
+) {
+  val currentObject = reportedObjectViewModel.selectedObject.value
```
You might want to use a `by reportedObjectViewModel.selectedObject.collectAsState()` here. The linter is here to help you and suppressing it for an error like this is not really recommended

LouisJamesVasseur replied:
I'll give it a try

matttsss reviewed (changes requested):
Alright... These are just a "few" indications for what needs to be worked on since this is still officially a draft. The main issue being the presence of Firebase in both ViewModels and this is quite a hard pill to swallow for me especially since I implemented DI to avoid this problem. (You also may find testing easier also if there is no Firebase in the ViewModels). Of course, we could look for solutions together if you are stuck on this when we meet later.
I haven't had a deep look into UI yet, since I had quite a lot to say on the model side, I'll make sure to check on it once you have addressed my issues.

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -31,21 +30,39 @@ class ReviewViewModel(
 
   fun selectReview(review: Review) {
     _selectedReview.value = review
+    loadReportsForSelectedReview()
```
selectReview is called in AdminScreen which is what it's intended for because of the loadReportsForSelectedReview().
But it's also called in the AllReviewsScreen. Im not sure why AllReviewsScreen would need to loadReportsForSelectedReview() ? Tell me if i misunderstood something, but it just seems to me like an unnecessary call, but it's not a big deal.

AlexandreBeeDeeBop replied:
So we could have like 2 different functions, one which selects and load reports for admin, and one which selects simply.

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -31,21 +30,39 @@ class ReviewViewModel(
     return reviewRepository.getNewUid()
   }
 
+  fun clearSelectedReview() {
```
Not used anywhere right now. 
Will it be for the onClick of the to-be-added delete object button that is shown on your PR description's image ?

LouisJamesVasseur replied:
I'm keeping it "in the back pocket" because I'll need it soon in some contexts

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:
```diff
@@ -83,41 +100,81 @@ class ReviewViewModel(
+                  reportUID = report.uid,
+                  nbOfTimesReported = selectedReview.nbReports + 1,
+                  nbOfTimesMaxSeverityReported =
+                      if (report.reason.severity == 3) selectedReview.nbMaxSeverityReports + 1
```
Same here, use MAX_SEVERITY.

LouisJamesVasseur replied:
Thanks!

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/report/ReportedObjectViewModel.kt:
```diff
@@ -0,0 +1,52 @@
+    _selectedObject.value = reportedObject
+  }
+
+  fun clearSelectedObject() {
```
Same here, this function is never used. Maybe it's for later then it's fine i guess...

LouisJamesVasseur replied:
Yep, I'll need it soon!

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/ui/report/ReviewReportScreen.kt:
```diff
@@ -70,18 +70,16 @@ fun ReviewReportScreen(
+            review = reviewId!!,
+            description = reportDescription.value)
+    reviewViewModel.addReport(report, userViewModel.currentUser.value!!)
+    Toast.makeText(context, "Report added", Toast.LENGTH_SHORT).show()
```
Put that toast message in string.xml as usual

LouisJamesVasseur replied:
Ahhh I missed one!

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/ui/report/ViewReportsScreen.kt:
```diff
@@ -0,0 +1,167 @@
+                  style = MaterialTheme.typography.bodyLarge)
+            }
+      },
+      floatingActionButtonPosition = FabPosition.End) { paddingValues ->
```
This can be simplified to avoid the repetition, especially if reports.isEmpty is true : 

`floatingActionButtonPosition = FabPosition.End) { paddingValues ->
    Box(modifier = Modifier.padding(paddingValues)) {
        val reports = when (selType) {
            ReportedObjectType.PARKING -> parkingViewModel.selectedParkingReports.value
            ReportedObjectType.REVIEW -> reviewViewModel.selectedReviewReports.value
        }

        Column(modifier = Modifier.fillMaxSize().padding(16.dp)) {
            if (reports.isEmpty()) {
                Text(
                    text = stringResource(R.string.no_reports_available),
                    style = MaterialTheme.typography.bodyMedium,
                    modifier = Modifier.align(Alignment.CenterHorizontally)
                )
            } else {
                LazyColumn(
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(vertical = 8.dp)
                ) {
                    items(reports) { report ->
                        val reportType = when (selType) {
                            ReportedObjectType.PARKING -> Report.Parking(report)
                            ReportedObjectType.REVIEW -> Report.Review(report)
                        }
                        ReportCard(reportType)
                    }
                }
            }
        }
    }
}
`

AlexandreBeeDeeBop replied:
Or if this works, even better :

```
floatingActionButtonPosition = FabPosition.End) { paddingValues ->
    Box(modifier = Modifier.padding(paddingValues)) {
        val reportCards: List<Report> = when (selType) {
            ReportedObjectType.PARKING -> 
                parkingViewModel.selectedParkingReports.value.map { Report.Parking(it) }
            ReportedObjectType.REVIEW -> 
                reviewViewModel.selectedReviewReports.value.map { Report.Review(it) }
        }

        Column(modifier = Modifier.fillMaxSize().padding(16.dp)) {
            if (reportCards.isEmpty()) {
                Text(
                    text = stringResource(R.string.no_reports_available),
                    style = MaterialTheme.typography.bodyMedium,
                    modifier = Modifier.align(Alignment.CenterHorizontally)
                )
            } else {
                LazyColumn(
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(vertical = 8.dp)
                ) {
                    items(reportCards) { reportCard ->
                        ReportCard(reportCard)
                    }
                }
            }
        }
    }
}

```

AlexandreBeeDeeBop reviewed (commented):
Once again a massive PR. Good job ! 
(And you know what else is massive ? The looooooowww taper fade meme)

 I know this is still technically a draft, but make sure to fix all those little issues I noticed. And Mattéo raises some good points too !

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingRepositoryFirestore.kt:
```diff
@@ -127,6 +127,26 @@ class ParkingRepositoryFirestore @Inject constructor(private val db: FirebaseFir
+      onSuccess: (List<ParkingReport>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    FirebaseFirestore.getInstance()
```
Will do, thanks!

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -332,40 +360,72 @@ class ParkingViewModel(
+
+          reportedObjectRepository.getDocumentReferenceByObjectUID(
+              objectUID = selectedParking.uid,
+              onSuccess = { documentRef ->
```
Note that we're in ParkingViewModel here. This deals with Reports (which are a subcollection of a Parking) and not an object, which is a separate thing entirely. You're not the first to get confused by this, may change the names in a future PR.
I didn't realize that making the changes inside the ViewModel instead of the Firestore made so much of a difference here, though it makes sense in retrospect. Shouldn't be too difficult a fix. Apologies for this and thanks for bringing to my attention

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -89,6 +88,15 @@ class ParkingViewModel(
   }
 
+  fun deleteParkingByUid(uid: String) {
+    parkingRepository.deleteParkingById(uid, {}, {})
```
Good call!

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -40,7 +39,7 @@ const val PARKING_MAX_SIDE_LENGTH = 50.0
 class ParkingViewModel(
     private val imageRepository: ImageRepository,
-    private val parkingRepository: ParkingRepository,
+    val parkingRepository: ParkingRepository,
```
I think I'm using it in a place I shouldn't be. I planned to remake it private before I fully opened the PR. Will do now

LouisJamesVasseur commented:
Thanks for all your comments to you both! I've spent a few hours checking everything and addressing all your comments and I'm now happy enough with my PR to finally open it

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingRepositoryFirestore.kt:
```diff
@@ -120,11 +121,53 @@ class ParkingRepositoryFirestore @Inject constructor(private val db: FirebaseFir
+            batch
+                .commit()
+                .addOnSuccessListener {
+                  if (querySnapshot.size() < 500) {
```
what's happening here ?

LouisJamesVasseur replied:
This is the condition that will allow us to verify that the subcollection is succesfully cleared: once the "reports" are indeed cleared, we can safely delete the parent collection

AlexandreBeeDeeBop commented on app/src/main/java/com/github/se/cyrcle/model/report/ReportedObjectRepository.kt:
```diff
@@ -61,6 +59,7 @@ interface ReportedObjectRepository {
   fun getObjectUID(
       objectUID: String,
       reportedObject: ReportedObject,
+      shouldAddIfNotExist: Boolean, // New parameter
```
useless comment

matttsss reviewed (changes requested):
Ok, a lot of small points, one major thing thought is the `ReportedObjectRepository::getObjectUID`, I believe this function has two distinct functionalities that can be separated into their own functions...

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/CyrcleNavHost.kt:
```diff
@@ -76,6 +80,7 @@ fun CyrcleNavHost(
         ReviewReportScreen(navigationActions, userViewModel, reviewViewModel)
       }
     }
+    navigation(startDestination = Screen.MAP, route = Route.MAP) {}
```
Think it's just a ktfmtFormat thing? I didn't add that

matttsss replied:
Right, but we agree this line should not exist right?

LouisJamesVasseur replied:
I'll check

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingRepositoryFirestore.kt:
```diff
@@ -120,11 +121,72 @@ class ParkingRepositoryFirestore @Inject constructor(private val db: FirebaseFir
+    val reportsCollectionRef = parkingDocRef.collection("reports")
+
+    // Recursive function to delete all documents in a collection
+    fun deleteSubcollection(
```
Me too! And trust me I wish it did!

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/ui/profile/ViewProfileScreen.kt:
```diff
@@ -83,6 +84,17 @@ fun ViewProfileScreen(
+        FloatingActionButton(
+            onClick = {
+              navigationActions.navigateTo(Screen.ADMIN)
+              Toast.makeText(context, "Admin FAB clicked!", Toast.LENGTH_SHORT).show()
```
Oh this one was meant to be outright removed haha

matttsss replied:
Trueeee I noticed that when I tested the UI on the phone and I was surprised about that toast xD

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/ui/report/AdminScreen.kt:
```diff
@@ -0,0 +1,312 @@
+}
+
+@Composable
+fun FilterSection(
```
We should! Issue for next Sprint, I say

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/ui/report/AdminScreen.kt:
```diff
@@ -0,0 +1,312 @@
+    parkingViewModel: ParkingViewModel,
+    reviewViewModel: ReviewViewModel
+) {
+  reportedObjectViewModel.fetchAllReportedObjects()
```
Good point

LouisJamesVasseur commented on app/src/test/java/com/github/se/cyrcle/model/report/ReportedObjectRepositoryFirestoreTest.kt:
```diff
@@ -73,6 +72,27 @@ class ReportedObjectRepositoryFirestoreTest {
+        onSuccess = { fail("Expected failure but got success") },
+        onFailure = { onFailureCalled = true })
+    taskCompletionSource.setResult(mockQuerySnapshot)
+    `when`(mockQuerySnapshot.isEmpty).thenReturn(true)
```
Well... if it works...

LouisJamesVasseur commented on app/src/androidTest/java/com/github/se/cyrcle/ui/report/AdminScreenTest.kt:
```diff
@@ -0,0 +1,146 @@
+import org.junit.Test
+import org.mockito.Mockito.mock
+
+class AdminScreenTest {
```
Good catch, missed that

AlexandreBeeDeeBop reviewed (approved):
All my comments have been adressed, now good luck with Mattéo but that's all for me.

LouisJamesVasseur commented on app/src/androidTest/java/com/github/se/cyrcle/di/mocks/MockReportedObjectsRepository.kt:
```diff
@@ -40,16 +27,34 @@ class MockReportedObjectRepository @Inject constructor() : ReportedObjectReposit
+      // Check if the object already exists to avoid duplicates
+      val existingObject = reportedObjects.find { it.reportUID == reportedObject.reportUID }
+      if (existingObject != null) {
+        throw IllegalArgumentException("ReportedObject with the same ReportUID already exists.")
```
good point... but as you mentioned it's only for tests. Isn't it better to see right away when an issue like this appears in tests?

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/report/ReportedObjectRepository.kt:
```diff
@@ -46,9 +56,26 @@ interface ReportedObjectRepository {
       objectUID: String,
-      onSuccess: (List<ReportedObject>) -> Unit,
+      reportedObject: ReportedObject,
+      shouldAddIfNotExist: Boolean, // New parameter
```
I've rolled out a change to separate this at your request

LouisJamesVasseur commented on app/src/androidTest/java/com/github/se/cyrcle/ui/report/ViewReportsScreenTest.kt:
```diff
@@ -0,0 +1,123 @@
+
+    val parkingReport = ParkingReport()
+    parkingViewModel.selectParking(TestInstancesParking.parking1)
+    composeTestRule.onRoot(useUnmergedTree = true).printToLog("ViewReportsScreen")
```
Prints out all the hierarchy of what's visible on the screen w/ test tags and everything. I recommend it's great!

matttsss reviewed (approved):
Thank you verry much for your changes! It looks good to me now!

LouisJamesVasseur commented:
Thanks for all the comments!
